tests/services/logging.service.spec.ts

```diff
@@ -1,64 +1,160 @@
+        loggingService.startSession();
+        loggingService.reset();
+      }
+      loggingService.startSession();
+      const actualLog = loggingService.getCachedLog();
+      let expectedLog = '';
+      for (let i = 0; i < loggingService.LOG_COUNT_LIMIT; i += 1) {
+        if (i === loggingService.LOG_COUNT_LIMIT - 1) {
+          expectedLog += headerLog;
+        } else {
+          expectedLog += `${headerLog}${sessionSeparator}`;
+        }
+      }
+      expect(actualLog).toEqual(expectedLog);
     });
+  });
 
-    it('should return additional logs if cache contains existing log', () => {
-      let logCounter = 1;
+  describe('.reset()', () => {
+    it('should do nothing if no session is ongoing', () => {
+      loggingService.reset();
+      const actualLog = loggingService.getCachedLog();
+      expect(actualLog).toBeNull();
+    });
 
-      while (logCounter < loggingService.LOG_COUNT_LIMIT) {
-        expect(getFilteredLogCount(repeatOldLogStartHeader(logCounter), logHeaderFilter)).toEqual(logCounter + 1);
-        logCounter += 1;
+    it('should not tamper with existing log histories', () => {
+      let expectedLog = headerLog;
+      for (let i = 0; i < loggingService.LOG_COUNT_LIMIT + 1; i += 1) {
+        loggingService.startSession();
+        loggingService.reset();
+        const actualLog = loggingService.getCachedLog();
+        expect(actualLog).toEqual(expectedLog);
+        if (i < loggingService.LOG_COUNT_LIMIT - 1) {
+          expectedLog += `${sessionSeparator}${headerLog}`;
+        }
       }
     });
+  });
 
-    it('should return updated log if log in cache contains max number of sessions', () => {
-      // Number of logs must stay the same
-      expect(getFilteredLogCount(repeatOldLogStartHeader(loggingService.LOG_COUNT_LIMIT), logHeaderFilter)).toEqual(
-        loggingService.LOG_COUNT_LIMIT
-      );
+  describe('adding logs', () => {
+    it('should successfully add info logs', () => {
+      testAddLog(infoLogMessage);
+    });
 
-      // Number of Old Logs must be reduced by 1 (To make way for the new session log)
-      expect(getFilteredLogCount(repeatOldLogStartHeader(loggingService.LOG_COUNT_LIMIT), oldLogFilter)).toEqual(
-        loggingService.LOG_COUNT_LIMIT - 1
-      );
+    it('should successfully add error logs', () => {
+      testAddLog(errorLogMessage);
     });
 
-    it('should return trimmed and updated log if log in cache exceeds max number of sessions', () => {
-      const exceededSessionCount = loggingService.LOG_COUNT_LIMIT + 10; // Arbitrary Exceed Count
+    it('should successfully add warn logs', () => {
+      testAddLog(warnLogMessage);
+    });
 
-      // Number of logs must be at max number
-      expect(getFilteredLogCount(repeatOldLogStartHeader(exceededSessionCount), logHeaderFilter)).toEqual(loggingService.LOG_COUNT_LIMIT);
+    it('should successfully add debug logs', () => {
+      testAddLog(debugLogMessage);
+    });
+  });
```

Comment:
Since `testAddLog` runs `loggingService.info` for all cases, perhaps it would be better to pass the logging function (eg `loggingService.debug/error/warn`) as a second argument.
Even so, all paths call `updateLogs`, which does not improve code coverage significantly. Inputs are also of the same equivalence class. Perhaps we can remove these tests altogether. What do the rest think? @CATcher-org/2021-devs

Reply:
@dingyuchen 
Should we test the `updateLogs` function instead?
I've made the `updateLogs` method `private` since based on my understanding, we're supposed to call these 4 methods to log stuffs.

Reply:
I think making `updateLogs` private is the right call, but I would argue that testing _one_ of `loggingService.(info|debug|error|warn)` is sufficient
ie just `testAddLog` once